identify/identify.py

```diff
@@ -61,15 +62,27 @@ def tags_from_path(path: str) -> Set[str]:
         tags.add(NON_EXECUTABLE)
 
     # As an optimization, if we're able to read tags from the filename, then we
-    # don't peek at the file contents.
+    # don't peek at the file contents unless the file extension requires it
     t = tags_from_filename(os.path.basename(path))
     if len(t) > 0:
         tags.update(t)
-    else:
-        if executable:
-            shebang = parse_shebang_from_file(path)
-            if len(shebang) > 0:
-                tags.update(tags_from_interpreter(shebang[0]))
+
+    ext = os.path.splitext(os.path.split(path)[-1])[-1].lstrip('.')
```

Comment:
I don't love that this introduces a need to parse the extension on all files.. but I don't see a way around that when the logic to determine the extra requirement is by extension. Thoughts?